### DISCUSSION TITLE
[Idea] Working with strings on the stack

### DISCUSSION THREAD

**Discussion by Ultrahead** [dotnet/csharplang#2092] [Language Ideas]
Recent versions of C# have introduced important features that improve performance and although there is [more to do](https://github.com/dotnet/csharplang/issues/1986) in such regard, imho, there is one key subject that must be addressed somewhere in the 8.x cycle: strings.

When you operate with strings, with the exception of a few cases, the output in general ends up generating a memory allocation on the heap, and therefore, garbage. For cases, for example, when you need to persist that output on a member field this is expected, but I believe that this norm should not apply for those cases where you operate with strings, temporarily. For the latter, the output of a string operation could be allocated somehow on the stack first, instead (and only moved/copied to the heap when the dev expresses such intent, say, as said before, assigning the output to a member field).

Expert devs may design and implement workarounds to avoid garbage, but such solutions usually may involve unsafe code and also may not be broadly applicable straightaway.

Applications that need every drop of performance they can get, like videogames, would benefit a lot from the possibility of allocating "strings" on the stack, temporarily. 

Picture this scenario (I'm using the example method from the docs) ...:

```csharp
void TestMethodOnTheHeap()
{
   string userName = "<Type your name here>";
   string dateString = DateTime.Today.ToShortDateString();

   // Use the + and += operators for one-time concatenations.
   string str = "Hello " + userName + ". Today is " + dateString + ".";
   System.Console.WriteLine(str);

   str += " How are you today?";
   System.Console.WriteLine(str);
}
```

... now replace it with something like the following ...:

```csharp
void TestMethodOnTheStack()
{
   stackstring userName = "<Type your name here>";
   stackstring dateString = DateTime.Today.ToShortDateString();

   // Use the + and += operators for one-time concatenations.
   stackstring str = "Hello " + userName + ". Today is " + dateString + ".";
   System.Console.WriteLine(str);

   str += " How are you today?";
   System.Console.WriteLine(str);
}
```

... provided that unions and WriteLine do support stackstring elements with no hidden strings, there should be no allocations on the heap. No garbage!

I do not know whether having a general stackstring is the solution here and maybe a different route should be followed, like integers. Say, string32, string64, string128, string256, string512, ..., string4096 (in bytes).

```csharp
void TestMethodOnTheStack()
{
   string16 userName = "<Type your name here>";
   string32 dateString = DateTime.Today.ToShortDateString();

   // Use the + and += operators for one-time concatenations.
   string128 str = "Hello " + userName + ". Today is " + dateString + ".";
   System.Console.WriteLine(str);

   str += " How are you today?";
   System.Console.WriteLine(str);
}
```

I do not know whether this route is a solution either, but I guess that now that we have structures like Span<T> that make use of CLR intrinsic elements, something could be achieved on this regard bringing a new feature that again help us boost performance with C#.

So I am hereby posting this idea so you can chime in and post how can we achieve something like this.

**Comment by yaakov-h** [comment]
```csharp
stackstring dateString = DateTime.Today.ToShortDateString();
```

A method returning a value allocated on the stack seems like either an impossibility or an inevitable disaster.

**Comment by PathogenDavid** [comment]
There's currently work being done to add general stack allocation support to CoreCLR:

* [Design document](https://github.com/dotnet/coreclr/blob/c0c51f83e56eb37bb5a257c0e82ca6676894d6c1/Documentation/design-docs/object-stack-allocation.md) ([PR with discussion](https://github.com/dotnet/coreclr/pull/20251))
* https://github.com/dotnet/coreclr/issues/1784 (Continued from https://github.com/dotnet/roslyn/issues/2104)
* [Prototype modifications to Roslyn and CoreCLR for explicit stack allocation](http://xoofx.com/blog/2015/10/08/stackalloc-for-class-with-roslyn-and-coreclr/#the-transient-variable-modifier) by Alexandre Mutel.

I think general/automatic stack allocation of small objects is probably the better route to go since it Just Works™.

----------

Without that, I think you're probably best of using `Span<char>` everywhere. It's not substantially different from what you're describing.

Unfortunately one issue you're going to run into with either solution is methods that can't handle `Span<char>`/`stackstring`. Unless a special method for `DateTime.ToShortDateString` is implemented, you're still going to be allocating on the heap. If object stack allocation is implemented *and* the Jit inlines `DateTime.ToShortDateString` (otherwise the string will get allocated in the stack frame for `ToShortDateString`), then this issue goes away.

--------------------------

> Applications that need every drop of performance they can get, like videogames, would benefit a lot from the possibility of allocating "strings" on the stack, temporarily.

IMO, we're better off with custom allocation in general for real-time situations. It'd be better to allocate the strings in a separate stack allocator that is nuked every frame. There's a proposal out there somewhere for allowing you to specify a custom allocator when allocating objects, but I can't seem to find it.

Local GC might be of interest for something like this:

* https://github.com/dotnet/coreclr/projects/3
* https://github.com/dotnet/coreclr/issues/11518
* http://tooslowexception.com/zero-garbage-collector-for-net-core/

**Comment by The-Futurist** [comment]
This is an interesting area, I've often wanted inline strings (for example structs that contain strings yet contain no reference types) and although these were implemented using fixed buffers the code we wrote was ugly.

So I'd like to propose a new type a string with a declared max length and possibly a new inline keyword.

Because we're declaring a max length the memory for the string can be allocated at compile time. There really is no need to insist that all strings must have unbounded capacity, there are many scenarios where we know or want to impose a max length yet C# offers nothing in this area.

**Comment by HaloFour** [comment]
Or maybe the compiler could support automatically converting string literals into `Span<char>`/`ReadOnlySpan<char>`?

```cs
ReadOnlySpan<char> span = "Hello World";
// equivalent to
ReadOnlySpan<char> span = stackalloc char[] { 'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd' };
```

#1947

**Comment by Ultrahead** [comment]
At first I thought of something like that, I must admit it, but that is more like syntactic sugar than a perf benefit. Plus working with literals is relatively cheap. I am thinking of a general solution that deal with more cases than just literals and that it can be also passed and returned to and from methods as parameters.

**Comment by HaloFour** [comment]
> and that it can be also passed and returned to and from methods as parameters.

If the content of the string is on the stack itself then you will be (necessarily) very limited in how that string can be passed around.  The rules would apply like they do for `ref struct` or `ref`s in general.  You couldn't create one in a method and return it.  You couldn't assign it to a field on the heap.  In short they'd be kind of useless.  And, as you mentioned, there really wouldn't be any perf benefit.

**Comment by Ultrahead** [comment]
I never said there would be no perf benefit. Plus, the idea is to find a safe way to achieve it. Not to say "It cannot be done" just because.

**Comment by HaloFour** [comment]
> Plus, the idea is to find a safe way to achieve it. Not to say "It cannot be done" just because.

The safe way is through `ref struct` types like `Span<char>` or `ReadOnlySpan<char>`, but it's the necessary limitations that make them relatively difficult to work with.  I don't see how you'd accomplish it any other way, nor do I see how you'd be able to treat this stack-allocated string as if it was a normal `System.String` as that would immediately break the rules that keeps those types safe.

I think the answer here is to have good APIs for working with `Span<char>`, maybe a `ref struct` version of `StringBuilder` where the consumer preallocates a block of working stack space.

**Comment by tannergooding** [comment]
> I think the answer here is to have good APIs for working with Span<char>, maybe a ref struct version of StringBuilder where the consumer preallocates a block of working stack space.

So `ValueStringBuilder`: https://github.com/dotnet/corefx/issues/28379

**Comment by HaloFour** [comment]
@tannergooding 

Exactly, and maybe for APIs like that it's worth having the compiler allow initialization of `Span<char>` from a string literal so that methods like `Append` and `Insert` can be easy to use.

**Comment by tannergooding** [comment]
> the compiler allow initialization of Span<char> from a string literal

I think it would effectively be just syntax sugar for the following code (which you can use as a workaround today):
```C#
ReadOnlySpan<char> ros = "Hello, World!".AsSpan();
Span<char> s = stackalloc char[ros.Length];
ros.CopyTo(s);
```

**Comment by HaloFour** [comment]
@tannergooding 

Wouldn't `stackalloc char[] { ... }` be a lot faster?

**Comment by tannergooding** [comment]
I don't believe so.

For `stackalloc char[] { ... }` the compiler currently initializes each element of the span individually.

While, for the `ros.AsSpan(); and ros.CopyTo()` case, we get a ROSpan directly from the string (zero-cost), we allocate a new buffer (something we have to do anyways), and do an efficient block copy of the data into the allocated buffer.

The latter is very similar to the `InitializeArray` optimization the compiler does elsewhere, but without having to also copy to an intermediate `char[]` first:
https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGA3AWACgUBmLAJgwGEMBvIjRrMlAFgwFlsAKASjoZMhAZQAOEAHYAeAMYALCAgB8GKBgC8qmBBkBrCABsDAexkZ5igNoBdOhgDkACXsAaB3FcODn+97f3jHxcfB38AdR9A/wQfPwcAEx8AQnsMAF8CQiE0wUZc5ix2DnI+ASyhJgAlOAh4gHkJAwBPMUlZBWUMBGM1TQAiRzgjYzcw4wQDeKS+gDoAQShWiT5MiqYl9sUVXq0dfWGzCwRLbqgZgBk4CQBzGDlrVbWunpmqY1EmgBVjbiheR8YOXKGHyogQAEsAG4QeBYbAANi6NXixkaTXMHRsGHO4PgCEMcwQ+PRmgkcAA7hirLZaA5nP4PP44r5IkEQj4Iv4og4YkyfIl/Cl0gCQcCMGCoTC4HDEdVag1mhsjiocXjDBoVbjEASiRAmiL8qRCpwSHx8vQxSJxNJlaoNLs9IYTIdMartQYLldbvcRUI3fiPW8Pt9fv98kC0kA

**Comment by YairHalberstadt** [comment]
Wouldn't the fastest be to initialise a span with the correct size, and insert each character in its correct position?

**Comment by YairHalberstadt** [comment]
Or even to use Unsafe.As to treat the ReadOnlySpan as a Span?

**Comment by tannergooding** [comment]
> Wouldn't the fastest be to initialise a span with the correct size, and insert each character in its correct position

That is what the following does. But it can do it with an efficient block-copy, rather than element-by-element initialization:
```
ReadOnlySpan<char> ros = "Hello, World!".AsSpan();
Span<char> s = stackalloc char[ros.Length];
ros.CopyTo(s);
```

> Or even to use Unsafe.As to treat the ReadOnlySpan as a Span?

This is the fastest, as it avoids the need to copy (you can just reference the string literal directly from memory). However, it is also dangerous and can break assumptions the runtime makes around string interning if you modify the backing data.

**Comment by YairHalberstadt** [comment]
I meant not to create a string in the first place, but to initialise the span with the characters.

**Comment by tannergooding** [comment]
> I meant not to create a string in the first place, but to initialise the span with the characters.

String literals don't have any runtime cost. They are emitted into the user string table (`#US`) in the PE file as part of the C# compilation process. The runtime then references them directly from memory. This is why you can do things like lock a string and have it impact other app-domains that are sharing the same assembly reference, or why the following program works the way it does:
```csharp
unsafe static void Main(string[] args)
{
    var s = "Hello, World!";

    Console.WriteLine(s);               // Hello, World!

    fixed (char* pS = s)
    {
        pS[0] = 'J';
    }

    Console.WriteLine(s);               // Jello, World!

    Console.WriteLine("Hello, World!"); // Jello, World!
}
```

Non-literal strings (such as values produced from `string.Format`) do have runtime cost to create the string, but once you have it, it is more efficient to just copy its data.

**Comment by YairHalberstadt** [comment]
Got it, thanks

**Comment by svick** [comment]
Since `stackstring` or `stringN` can't really work as proposed (they can't be returned from another method) and since `Span`, including APIs to support it, exist, I thought it would be nice to see how the example code would look like using those. My attempt (including a trivial `ValueStringBuilder`):

```c#
void TestMethodUsingSpan()
{
    ReadOnlySpan<char> userName = "<Type your name here>";
    Span<char> dateString = stackalloc char[32];
    DateTime.Today.TryFormat(dateString, out int charsWritten, "d");
    dateString = dateString.Slice(0, charsWritten);

    Span<char> buffer = stackalloc char[128];
    var sb = new ValueStringBuilder(buffer);

    sb.Append("Hello ");
    sb.Append(userName);
    sb.Append(". Today is ");
    sb.Append(dateString);
    sb.Append(".");
    Console.Out.WriteLine(sb.AsSpan());

    sb.Append(" How are you today?");
    Console.Out.WriteLine(sb.AsSpan());
}

ref struct ValueStringBuilder
{
    Span<char> buffer;
    int position;

    public ValueStringBuilder(Span<char> initialBuffer)
    {
        buffer = initialBuffer;
        position = 0;
    }

    public ReadOnlySpan<char> AsSpan() => buffer.Slice(0, position);

    public void Append(ReadOnlySpan<char> value)
    {
        value.CopyTo(buffer.Slice(position));
        position += value.Length;
    }
}
```

It's not particularly nice code and I believe it will actually allocate today (because `Console.Out` does not override `WriteLine(ReadOnlySpan<char>)` and the default implementation has to allocate).

But the point is that this is almost possible today (`ValueStringBuilder` being the only truly missing piece). And because this is about performance critical code, some ceremony is probably acceptable. If there are improvements to be made, they would likely be in the form of improving the available APIs, or maybe even language features targeting specific parts of the above code (span string interpolation? easier way to stack allocate span of the right size for the `TryFormat` method?), not adding yet another string type to C# (or even a family of them).

**Comment by Ultrahead** [comment]
>  not adding yet another string type to C# (or even a family of them ...

I understand, and I would also want to avoid that. The problem is that instances of classes in C# are always allocated on the heap. Strings are kind of an exception only when you reference literals. So perhaps an alternative route would be to modify the CLR, as an exception, for strings so that they can be also allocated on the stack. Perhaps this could be achieved with stackalloc ...

```csharp
stackalloc string myString = stackalloc $"Hello {$name}!";

stackalloc string myString = $"Hello {$name}!";

string myString = stackalloc $"Hello {$name}!";
```

... given that string is a special (immutable) class.

**Comment by CyrusNajmabadi** [comment]
@Ultrahead You can already accomplish effectively that by using fixed-size-buffers and `Span<char>`.  Importantly, `Span<char>` has huge restrictions on how you can use it so that you cannot violate memory safety.  i.e. someone woudl not be able to accidentally have a pointer to the chars on your stack which might then go away when your stack frame disappears.  'string' has no such safety.  imagine writing what you wrote and then passing that 'string' anywhere else.  It could immediately become corrupt once your stack went away.

For that reason, you should use `Span<char>` and then have all the safety around it.  There are already lots of helpers to enable you to to common sorts of operations on a `Span<char>` (for example, finding substrings in it and whatnot).  And, if you find some operations missing, you could always write them yourself, or suggest corefx provide it.

**Comment by Ultrahead** [comment]
> It could immediately become corrupt once your stack went away.

I guess that by "anywhere else" you mean the heap. In that case, it should be passed by copy.

> There are already lots of helpers to enable you to to common sorts of operations on a Span<char> ...

Are there any hidden allocations in such operations? The idea here is to avoid any allocation on the heap, unless intended by the dev.

**Comment by CyrusNajmabadi** [comment]
> I guess that by "anywhere else" you mean the heap. In that case, it should be passed by copy.

That's effectively *everything* you want to do with a string.  Can you name a single thing you could do here that would not involve a copy?

**Comment by CyrusNajmabadi** [comment]
> Are there any hidden allocations in such operations? The idea here is to avoid any allocation on the heap, unless intended by the dev.

No.  There are no hidden allocations.

**Comment by Ultrahead** [comment]
> That's effectively everything you want to do with a string.

On the contrary, I want to avoid copies unless intended ... and allocations on the heap.

@PathogenDavid: I've just read your post. General stack allocation on the CLR is really interesting and fits what I intend with strings. Thanks for posting it.

**Comment by CyrusNajmabadi** [comment]
> On the contrary, I want to avoid copies unless intended ... and allocations on the heap.

I'm saying: anything you would do with this 'stack allocated string' would need to cause a copy.  How else could it work?  After all, basically anything you do with a string might end up leaking the string somewhere.

--

Or, in other words, i'm flipping this around on *you*.  If you want a stack-allocated-string you need to present information on how this would *actually* work.  How it woudl be safe.  How it would be efficient.  How it would prevent implicit allocations etc.

As mentioned already, there is already a solution for this: `Span<char>`.  'string' today is miles away from that.  So bridging that gap will require a lot of explanation on your part.

**Comment by Ultrahead** [comment]
> I'm saying: anything you would do with this 'stack allocated string' would need to cause a copy.

Will that apply when working with ref/out?

> After all, basically anything you do with a string might end up leaking the string somewhere.

The idea is to avoid exactly that.

> If you want a stack-allocated-string you need to present information on how this would actually work.

Automatic memory (C++). 

> So bridging that gap will require a lot of explanation on your part.

Please read the post of @PathogenDavid and follow the links for explanations (there is a nice one by @xoofx).

**Comment by HaloFour** [comment]
All of the CLR work to support stack allocation is 100% implementation details and hidden to the developer.  The JIT will do runtime analysis to ensure that these objects cannot escape the stack in which they are created and only then will it allow for the object to be created on the stack.  That escape analysis is critical as without it any stack allocation is inherently dangerous.  This is why `ref` types and `ref` locals/returns are so limited.  This is why you can't have a stack-allocated string that behaves like a normal `string`.

**Comment by Ultrahead** [comment]
I did not know about this article, which I recommend reading: http://xoofx.com/blog/2015/10/08/stackalloc-for-class-with-roslyn-and-coreclr/#the-transient-variable-modifier

**Comment by CyrusNajmabadi** [comment]
> Will that apply when working with ref/out?

Sure.  if you pass by ref, someone may still make a copy of it.  Nothing stops that.

> The idea is to avoid exactly that.

Understood.. but **you** are not explaining how that would work with System.Strings.  I get that you *want* this.  But it's not like that means it's possible.  If you want this, you need to expand more on how ths would actually work.

> Automatic memory (C++).

That's a non-answer.  C# is not C++, and it's not like it's going to absorb the entire C++ object/memory/copying model.  If you specifically want 'string' to work (and are not willing to work with the actual parts of the language designed to be stack-alloc only) then you really need to do more legwork to explain how that would actually work. i.e. what rules does hte language need for this.  How do these work with normal heap strings.  How does the runtime work.  etc. etc. etc.  

> Please read the post of @PathogenDavid and follow the links for explanations (there is a nice one by @xoofx).

I've read all of this.

**Comment by Ultrahead** [comment]
> I've read all of this.

If so, then you can find there all the answers you want me to expand, explain or do legwork, etc. etc. etc.

**Comment by CyrusNajmabadi** [comment]
> If so, then you can find there all the answers you want me to expand, explain or do legwork, etc. etc. etc.

Clearly not.  For example, you've stated you don't want implicit allocations happening without you knowing it. However, the work being done in this space is about *conservative* automatic escape analysis.  In such a world, this is just a behind-the-scenes perf optimization.  It does not give you the featureset you want, and you could easily run into situations where you thought you wouldn't allocate, but you would.  This might even change from runtime version to version as the impl changed.

**Comment by CyrusNajmabadi** [comment]
> On the contrary, I want to avoid copies unless intended ... and allocations on the heap.

@Ultrahead That was your desired outcome.  The links and discussions do not have a solution for that *outside of doing what i said*, namely, using `Span<char>`.  If you want to use `string` and have that behavior, then **you** are going to need to do some extra legwork, as there's no proposal or design currently (that i'm aware of) or anything in the links you've provided that shows how that would work.

**Comment by Ultrahead** [comment]
lol, you have a fixation with me ... "you", "you", "**you**" ... what's next? "**you!**"

> Clearly not. 

I don't agree. @xoofx did a great job, including an example with benchmarks:

>The stack version will run in 400 ms with 0 GC collect
> The heap version will run in 5000 ms with 100+ GC collect

Those numbers speak for them-selves.

**Comment by CyrusNajmabadi** [comment]
> I don't agree. @xoofx did a great job, including an example with benchmarks:

The experiment there is entirely incomplete, and has major caveats and holes around how it could work and be safe.  Unless we're reading different articles, then that's nothing more than a proof of concept with tons more that would be necessary to actually make things possible.

> Those numbers speak for them-selves.

They do nothing of the sort.  No one is questioning the value in being able to stack alloc things.  But having numbers doesn't magically make things possible.  It's simply to carve out a very narrow experiment, leave out lots of stuff (like actual safety), and then show positive numbers.  It's another thing to actually make it to provide a viable working model that meets the use case requirements that people have laid out.

> lol, you have a fixation with me 

Dude... you literally said you wanted something.  I've been pointing out how what you want is actually quite hard to provide.  Your responses that point to incomplete experiments, or flippantly stating that things should just be like C++, changes none of that.

At this point, i'm just going to assume you're not actually engaging in a discussion around the very thing you said you wanted.  So i'll leave hte discussion right where it started:

If you want a stack allocated string-like concept.  Then just use `Span<char>`.  It gives almost exactly the semantics you want.  It's on the stack.  Can't accidentally be copied anywhere.  Will not cause any implicit allocations.  And, it can be treated mostly like a string.  If you need a true string from it, then that's not hard to provide, and you will be in control of the allocation.  This will give similar numbers to what you've seen in these links, without needing any new language features or CLR changes.

If that's  good enough for you, great!  If not, you'll have to explain why not, what it is you actually want, and then *how* you would even go about getting it. Cheers!

**Comment by xoofx** [comment]
> The experiment there is entirely incomplete, and has major caveats and holes around how it could work and be safe. Unless we're reading different articles, then that's nothing more than a proof of concept with tons more that would be necessary to actually make things possible.

Certainly yes.

That being said, as someone suggested in one comment at the end of the article, the most prominent issue was for ensuring `this` to not escape but by tagging methods with the `transient` keyword, it should ensure that `this` doesn't escape, and you would not be able to call a method on a transient object that is not transient. If you think that this doesn't work, I would really be glad to hear about it, as well as If you have in mind other issues with some concrete examples, that would help!

**Comment by Ultrahead** [comment]
> If not, you'll have to explain why not, 

The language should be modified in a way that it always avoids hidden allocations on the heap when temporarily operating with strings. Example: $"Hello {name}!"; 

> what it is you actually want, 

Simplicity. Transient variables for reference types looks indeed promising to me. 

> and then how you would even go about getting it.

@xoofx has already posted lots of details on his article, including the "how". 

Cheers!

**Comment by CyrusNajmabadi** [comment]
> That being said, as someone suggested in one comment at the end of the article, the most prominent issue was for ensuring this to not escape but by tagging methods with the transient keyword, it should ensure that this doesn't escape, and you would not be able to call a method on a transient object that is not transient. If you think that this doesn't work, I would really be glad to hear about it, as well as If you have in mind other issues with some concrete examples, that would help!

Hi @xoofx!  Have really enjoyed your posts.  

It def seems like something akin to the rules for ref-structs would need to be enforced here.  And, it's basically unclear how this could ever help the use case  @Ultrahead cares about.  Specifically, strings wouldn't have these attributes (nor would all the methods out there in the world that took strings).  So he would effectively not be able to do anything with these stack-alloc'ed strings.

**Comment by HaloFour** [comment]
@Ultrahead 

> The language should be modified in a way

The language can't be modified in a way that isn't supported by the CLR, which has no concept of stack-allocation of reference types, let alone strings.  You'd need to start there.

**Comment by CyrusNajmabadi** [comment]
> @xoofx has already posted lots of details on his article, including the "how".

Could you point out specifically what you're referring to?  Even if @xoofx's 'transient' idea was accepted in the language, and if the CLR started doing special stuff around it.  That certainly wouldn't benefit you *at all* in the case where you said something like `transient string s = ....`.  Nor would it ever help you for hte `$"Hello {name}!"` case (which would involve formattable strings and definite heap allocations).

Basically, what i'm seeing is:

1. Do something
2. Wave hands
3. Strings are now allocated on the stack and have exactly the semantics you want.

There is literally no proposal out there afaict that are giving you what you want.  The coreclr work has been on automatic escape analysis.  That would not help your case as there is nothing in that preventing automatic implicit allocations (something you firmly stated were unacceptable).  The xoofx work has been about how future stuff might work, but which would require a huge amount of ecosystem opt-in to ensure things were ok.  Very specifically, it couldn't work with string due to basically everything under the sun expecting to be able to take in a string and do whatever it wants with it.  So 100% of the ecosystem would not be able to use your stack-alloc'ed strings.

This is the benefit of `Span<char>`.  it starts *today* as something that ensures you get stack-allocing.  And, anything that wants to work with it already knows that it has restrictions on how it can be used to ensure that nothing is violated.

The 'transient' restriction that xoofx has mentioned, are akin to the `ref struct` restrictions the language has already introduced.  Except that they can work today, instead of needing some way to retro-fit them onto existing types that are literally used in 100% of codebases in so many ways that would violate those constraints.

**Comment by xoofx** [comment]
> Specifically, strings wouldn't have these attributes (nor would all the methods out there in the world that took strings). So he would effectively not be able to do anything with these stack-alloc'ed strings.

Hi @CyrusNajmabadi! Sorry, got dragged into this discussion 😀 

I don't think that any strings method have a this pointer that can escape, so they could potentially have all their method being marked transient.... but yeah, it is a concept that is making ownership storage explicit, so, it would unlikely come to C# considering the amount of API that would have to be updated to reflect this to make it really usable.

Incidentally I'm revisiting this with a reboot of my Stark language experiment (but now based on C# Roslyn parser), and I'm going to take the opposite default:  you need to explicitly tag a method or parameter with the keyword `retainable` to allow the struct/class to be stored elsewhere than on the stack (member, static...etc.)

But overall, I understand how much it is difficult to change a language, there are so many dependencies and constraints that it is difficult to make a move without dragging breaking changes...

**Comment by xoofx** [comment]
> The 'transient' restriction that xoofx has mentioned, are akin to the ref struct restrictions the language has already introduced. 

Exactly, the difference is that it doesn't make a `ref` but it keeps the other constraint of `ref`, so it can work for class type or struct value directly. And I agree, in practice, it can be very painful (exactly like `ref` cannot be used in async patterns for example where an implicit capture is being done)

**Comment by CyrusNajmabadi** [comment]
> Incidentally I'm revisiting this with a reboot of my Stark language experiment (but now based on C# Roslyn parser), and I'm going to take the opposite default: you need to explicitly tag a method or parameter with the keyword retainable to allow the struct/class to be stored elsewhere than on the stack (member, static...etc.)

Nice!  That's like how i'd like to do a new language as well.  Make things as cheap/zero-cost-abstractiony as possible, then have ways to explicitly opt-into the costs.  It's been a bit of fresh air to get closer to that in the Rust/Go spaces (especially as the latter still has GC).  

I don't see any easy way to get from where C# is to there though given the massive ecosystem/history burden.  I think we'll just ahve to attack this by providing new features in the zero-cost space, and having stuff use that as appropriate.  i think the Span/ref work go a long way there, but it's clearly going to be a continual process.  I expect we'll need solutions for lambdas/await that play nicely in that space very soon.

**Comment by Ultrahead** [comment]
> Very specifically, it couldn't work with string due to basically everything under the sun expecting to be able to take in a string and do whatever it wants with it. So 100% of the ecosystem would not be able to use your stack-alloc'ed strings.

I agree that there's lot of operations that may require an update to fit the needs of a stack-allocated string. But I also believe the idea proposed by Alex is in the right path. And we should move step by step. Once transient vars (or similar feature) is included in the language&CLR, we could move to step 2:

```CSharp
transient var mystring = myField.ToString(); // Error if ToString method is not marked as transient

transient var mystring = transient myField.ToString(); // (2)
```

On (2) above, the compiler may attempt to create a transient extension method for ToString() and produce an error if it cannot achieve it.

```

**Comment by CyrusNajmabadi** [comment]
> And we should move step by step.

Isn't that what you can do today with `Span<char>`?  you can make your stack allocated strings, and have total contorl over them (ensuring they don't every allocate or do anything you don't want).  Then, at the point you want a string that has more capabilities, you choose the point at which to make the copy and then work with that.   The language and APIs already support this today, and they have what seems to be the semantics you want.

**Comment by xoofx** [comment]
> `transient var mystring = myField.ToString();`

My proposal didn't say that this was possible.  A method can return a transient only if it is coming from a transient parameter, otherwise it is not possible (exactly like ref). Otherwise, the only function able to return a transient string is the stackalloc operator.  I didn't read the top of this proposal, but yeah, this cannot work at all, a function cannot allocate on the stack and return it to the caller.

**Comment by arekbal** [comment]
```
ValueString32 mystring = myField.ToValueString32();
mystring.Append("I love ice cream".AsSpan());
```
That is what standard function call with returned value actually does... leaves value on the stack for caller to consume (ignoring registers and copies for sake of argument).
Simple, fast, and working. Requires one-time extra effort 'only'... 
Using such ValueStringXX structs with Spans is a breeze.
PS. Dynamic strings are way harder and less safe btw...

**Comment by HaloFour** [comment]
@arekbal

> That is what standard function call with returned value actually does... leaves value on the stack for caller to consume

No, that is not how the stack works.  A function cannot allocate space on the stack and then pass a reference to that space back to the caller.  As soon as that function returns that stack space is effectively reclaimed and will be overwritten by any other method call.

For a function to populate stack space with anything that will live after that function returns the calling function has to allocate the space and pass a reference.

For functions returning value types the value type must be copied into the stack space of the calling function, which requires the type be a predictable size.

Even in languages with arbitrary pointers you can't just return a pointer to space you allocate on the stack, not without expecting your program to fail shortly thereafter.

**Comment by HaloFour** [comment]
@arekbal

I see you mention using a different struct for each "string" size.  Yes, that would work, but it does involve making a bunch of separate and incompatible types, and you incur a copy of the "string" on return.

**Comment by arekbal** [comment]
@HaloFour 

> I see you mention using a different struct for each "string" size. Yes, that would work, but it does involve making a bunch of separate and incompatible types

Yeah, "we" had this discussion in somewhat similar subject and how to deal with those.

>and you incur a copy of the "string" on return

I just stupidly assumed RVO... so the way to use it efficiently today instead is through refs or outs. I guess.

Here is some cool article about RVO in C# but you guys know better than me.
http://blog.s-schoener.com/2018-03-24-rvo-csharp/

**Comment by arekbal** [comment]
@HaloFour Caller function can allocate stack space and pass an adress to it to called function. just what ref does. RVO would mean just that or having it as extra on the stack in frame of the caller for caller to cleanup. Stack is only about moving that pointer up and down over that memory. Depending on what calling convention you use or assembly you're willing to emit, in bare asm there is little stopping you from returning stack allocated space to the caller. It is up for the compiler to decide how to enable such operation so it would be safe.
